presto-main/src/main/java/com/facebook/presto/resourcemanager/DistributedResourceGroupInfoResource.java

```diff
@@ -0,0 +1,155 @@
+import static javax.ws.rs.core.Response.Status.NOT_FOUND;
+
+@Path("/v1/resourceGroupState")
+@RolesAllowed(ADMIN)
+public class DistributedResourceGroupInfoResource
+{
+    private static final Logger log = Logger.get(DistributedResourceGroupInfoResource.class);
+    private final InternalNodeManager internalNodeManager;
+    private final ListeningExecutorService executor;
+    private final HttpClient httpClient;
+    private final JsonCodec<ResourceGroupInfo> jsonCodec;
+
+    @Inject
+    public DistributedResourceGroupInfoResource(InternalNodeManager internalNodeManager,
+            @ForResourceManager ListeningExecutorService executor, @ForResourceManager HttpClient httpClient, JsonCodec<ResourceGroupInfo> jsonCodec)
+    {
+        this.internalNodeManager = requireNonNull(internalNodeManager, "internalNodeManager is null");
+        this.executor = requireNonNull(executor, "executor is null");
+        this.httpClient = requireNonNull(httpClient, "proxyHelper is null");
+        this.jsonCodec = requireNonNull(jsonCodec, "jsonCodec is null");
+    }
+
+    @GET
+    @Produces(MediaType.APPLICATION_JSON)
+    @Encoded
+    @Path("{resourceGroupId: .+}")
+    public void getResourceGroupInfos(
+            @PathParam("resourceGroupId") String resourceGroupIdString,
+            @HeaderParam(X_FORWARDED_PROTO) String xForwardedProto,
+            @Context UriInfo uriInfo,
+            @Context HttpServletRequest servletRequest,
+            @Suspended AsyncResponse asyncResponse)
+    {
+        if (isNullOrEmpty(resourceGroupIdString)) {
+            asyncResponse.resume(Response.status(NOT_FOUND).build());
+        }
+        try {
+            ImmutableList.Builder<ListenableFuture<ResourceGroupInfo>> queryStateInfoFutureBuilder = ImmutableList.builder();
+            for (InternalNode coordinator : internalNodeManager.getCoordinators()) {
+                queryStateInfoFutureBuilder.add(getResourceGroupInfoFromCoordinator(xForwardedProto, uriInfo, coordinator));
+            }
+            List<ListenableFuture<ResourceGroupInfo>> queryStateInfoFutureList = queryStateInfoFutureBuilder.build();
+            Futures.whenAllComplete(queryStateInfoFutureList).call(() -> {
+                try {
+                    ResourceGroupInfo aggregatedResourceGroupInfo = aggregateResourceGroupInfo(queryStateInfoFutureList);
+                    if (aggregatedResourceGroupInfo == null) {
+                        return asyncResponse.resume(Response.status(NOT_FOUND).build());
+                    }
+                    return asyncResponse.resume(Response.ok(aggregatedResourceGroupInfo).build());
+                }
+                catch (Exception ex) {
+                    log.error(ex, "Error in getting resource group info from one of the coordinators");
+                    return asyncResponse.resume(Response.serverError().entity(ex.getMessage()).build());
+                }
+            }, executor);
```

Comment:
I would write this more along these lines:

```java
            ListenableFuture<List<ResourceGroupInfo>> resourceGroupInfosFuture = Futures.allAsList(queryStateInfoFutureList);
            addSuccessCallback(resourceGroupInfosFuture, resourceGroupInfos -> {
                ResourceGroupInfo aggregatedResourceGroupInfo = buildResourceGroupInfo(resourceGroupInfos);
                if (aggregatedResourceGroupInfo == null) {
                    asyncResponse.resume(Response.status(NOT_FOUND).build());
                    return;
                }
                asyncResponse.resume(Response.ok(aggregatedResourceGroupInfo).build());
            });
            addExceptionCallback(resourceGroupInfosFuture, e -> {
                log.error(e, "Error in getting resource group info from one of the coordinators");
                asyncResponse.resume(Response.serverError().entity(e.getMessage()).build());
            });
```

Where `buildResourceGroupInfo` is a new method that takes in a list of `ResourceGroupInfo` and aggregates it in one shot.  We can also remove `AggregatedResourceGroupInfoBuilder` this way.

Reply:
One problem I see with this is, it does not handle the case where one coordinator throws 404 and others returns successful response. In this case, we still want to aggregate the successful response and return.

I think with moving the aggregation to inline method will complicate aggregation for subgroups, having a builder will avoids this complication as well as need of creating more objects.